### PR TITLE
Fix wrong status 404 when PATCH request didn't change anything

### LINKED ISSUE #1257: 404 when PATCH/update didn't change anything
Hello. I'm doing a PATCH to update a row that doesn't exist. There's a GitHub issue (https://github.com/PostgREST/postgrest/issues/182) and pull request/changelog entry from 2015 (https://github.com/PostgREST/postgrest/commit/709e70561f8b527998b8ed8ac0916e92b1a5e17b) that indicates a PATCH that didn't change any rows should return 404.

Right now in my tests, I'm getting `204` whether or not any rows were updated. Could it be that this "404 when no rows were updated" behavior got lost somewhere along the line?

I'm using `postgREST v0.5.1` and Postgres 9.

### PR DESCRIPTION
**Pull request by Qu4tro**
This is a simple attempt to fix the reported bug described in title.
So, citing #1257 and confirmed by this test in this PR:

> Right now in my tests, I'm getting 204 whether or not any rows were updated. 

This PR can be tweaked to take care of some of the failing tests, but like in the previous PR, I would argue that some of those failing tests should be tweaked as well.

Let me know what you think.

This resolves #1257 .

### CODE REVIEW

**Comment by Qu4tro** [comment]
Added a few changes:

Using a empty payload no longer touches the db. They keep returning the same status as before.
PATCHs whose conditions result in no records found (and therefore no records updated) return 404.

1º was a necessary condition for 2º, because we need to distinguish between, declaring no changes and not founding records to change, which return different status codes.

What do you think?

File: src/PostgREST/App.hs [hunk 1]
```diff
@@ -188,21 +188,47 @@ app dbStructure proc conf apiRequest =
...
-                          then status200
-                          else status204
-                return $ if iPreferRepresentation apiRequest == Full
-                  then responseLBS s [toHeader contentType, r] (toS body)
-                  else responseLBS s [r] ""
+
+                  contentRangeHeader = contentRangeH 0 (toInteger $ queryTotal-1)
+                        (toInteger <$> if shouldCount then Just queryTotal else Nothing)
+
+                  updateIsNoOp = case mutationDbRequest tSchema tName of
```
Comment (steve-chavez) [inline src/PostgREST/App.hs]: This data can be obtained with `pjKeys pJson`(see PayloadJSON in Types.hs), without passing through `mutateRequest`.

**Comment by steve-chavez** [comment]
> we need to distinguish between, declaring no changes and not founding records to change, which return different status codes.

Yes. This is correct :+1: 

However, the wildcard pattern matchs are a bit hard to follow. I think not hitting the database should not be a concern for now and the code can be a bit more concise(the added variables are good though). It could be something like:
```haskell
let
  contentRangeHeader = contentRangeH 0 (toInteger $ queryTotal - 1)
        (toInteger <$> if shouldCount then Just queryTotal else Nothing)
  minimalHeaders = [contentRangeHeader]
  fullHeaders = toHeader contentType : minimalHeaders
  updateIsNoOp = S.null $ pjKeys pJson
  status | queryTotal == 0 && not updateIsNoOp = status404
         | iPreferRepresentation apiRequest == Full = status200
         | otherwise = status204
return $ if iPreferRepresentation apiRequest == Full
  then responseLBS status fullHeaders (toS body)
  else responseLBS status minimalHeaders mempty
```

**Comment by Qu4tro** [comment]
A few more commits.

- `pjKeys`
- Simplified the case statement. I've kept it over the `if`s expression, to better handle the `CTSingularJSON` possibility.
- Made contentRangeH polymorphic, so we can reduce the number of `toInteger` calls.

**Comment by Qu4tro** [comment]
Additional commit for some whitespace. I think it reads better.

File: src/PostgREST/App.hs [hunk 2]
```diff
@@ -188,21 +188,31 @@ app dbStructure proc conf apiRequest =
...
+                         | iPreferRepresentation apiRequest == Full = status200
+                         | otherwise                                = status204
+
+              case (contentType, queryTotal, iPreferRepresentation apiRequest) of
+                (CTSingularJSON, 1, Full) ->
+                  return $ responseLBS status fullHeaders (toS body)
+
+                (CTSingularJSON, _, Full) ->
+                  do HT.condemn
+                     return $ singularityError 0
```
Comment (steve-chavez) [inline src/PostgREST/App.hs]: Previously we had:
```haskell
return $ singularityError (toInteger queryTotal)
```
It'd be better to keep the queryTotal instead of always returning 0.

I'm surprised this wasn't included in any test. Perhaps you could add one for making sure we don't miss this again.

File: src/PostgREST/App.hs [hunk 3]
```diff
@@ -381,7 +392,7 @@ rangeStatus lower upper (Just total)
   | (1 + upper - lower) < total = status206
   | otherwise               = status200
 
-contentRangeH :: Integer -> Integer -> Maybe Integer -> Header
+contentRangeH :: (Integral a, Show a) => a -> a -> Maybe a -> Header
```
Comment (steve-chavez) [inline src/PostgREST/App.hs]: Nice! :+1:

File: src/PostgREST/App.hs (hunk 2 above)
Comment (Qu4tro) [reply src/PostgREST/App.hs]: Ah, nice catch, thanks!

**Comment by Qu4tro** [comment]
Further commit that fixes the issue mentioned above. I've also hardcoded the tests, so it doesn't only check if it is an error, but which as well.

Curious thing happened on 
https://github.com/PostgREST/postgrest/pull/1272/files#diff-d36f1ad401857a200ca8d27e932681f8L64
Re-running the tests (without destroying db), yields different results (6 or 7).
Kept previous check instead, as the case its tested further down.

File: src/PostgREST/App.hs [hunk 4]
```diff
@@ -188,21 +187,30 @@ app dbStructure proc conf apiRequest =
...
+                  updateIsNoOp       = S.null $ pjKeys pJson
+                  contentRangeHeader = contentRangeH shouldCount 0 (queryTotal - 1) queryTotal
+                  minimalHeaders     = [contentRangeHeader]
+                  fullHeaders        = toHeader contentType : minimalHeaders
+
+                  status | queryTotal == 0 && not updateIsNoOp      = status404
+                         | iPreferRepresentation apiRequest == Full = status200
+                         | otherwise                                = status204
+
+              case (contentType, queryTotal, iPreferRepresentation apiRequest) of
```
Comment (steve-chavez) [inline src/PostgREST/App.hs]: queryTotal is only used by 1 match, so how about expressing this as:
```haskell
case (contentType, iPreferRepresentation apiRequest) of
  (CTSingularJSON, Full)
    | queryTotal == 1 -> return $ responseLBS status fullHeaders (toS body)
    | otherwise -> HT.condemn >> return (singularityError queryTotal)

  (_, Full) ->
    return $ responseLBS status fullHeaders (toS body)

  (_, _) ->
    return $ responseLBS status minimalHeaders mempty
```

File: src/PostgREST/App.hs [hunk 5]
```diff
@@ -322,7 +329,7 @@ app dbStructure proc conf apiRequest =
       rangeHeader queryTotal tableTotal =
         let lower = rangeOffset topLevelRange
             upper = lower + toInteger queryTotal - 1
-            contentRange = contentRangeH lower upper (toInteger <$> tableTotal)
+            contentRange = contentRangeH (isJust tableTotal) lower upper $ (toInteger . fromJust) tableTotal
```
Comment (steve-chavez) [inline src/PostgREST/App.hs]: We should [avoid partial functions](https://wiki.haskell.org/Avoiding_partial_functions) as much as we can. There are some of those in the codebase, but I've been removing them in any chance I get. Could you express this in other way and avoid `fromJust`?

Comment (Qu4tro) [reply src/PostgREST/App.hs]: Yes, I understand.

I've changed to this form, because it was shorter and `total` was never instantiated if `shouldCount` was false. 

Anyway, to avoid this form, which is I agree, is not ideal, I just reverted to the previous form.

File: src/PostgREST/App.hs (hunk 4 above)
Comment (Qu4tro) [reply src/PostgREST/App.hs]: Changed to suggested.

**Comment by steve-chavez** [comment]
@Qu4tro Thank you! Sorry the review took so long.
